src/interval.jl

```diff
@@ -351,6 +388,203 @@ true
 ≫(a, b) = greater_than_disjoint(a, b)
 # ≫̸(a, b) = !≫(a, b)
 
+###### Set-related Helpers #####
+
+# unbunch: represent one or more intervals as a sequence of endpoints and their indices into the original intervals
+endpoint_type(::Type{<:AbstractInterval{T}}) where T = Endpoint{T}
+endpoint_type(::Type{<:AbstractVector{T}}) where T <: AbstractInterval = endpoint_type(T)
+endpoint_type(x) = endpoint_type(typeof(x))
+
+leftpoint(x::AbstractInterval, ::Type{<:Endpoint}) = LeftEndpoint(x)
+rightpoint(x::AbstractInterval, ::Type{<:Endpoint}) = RightEndpoint(x)
+leftpoint(x::AbstractInterval, ::Type{<:HalfOpenEndpoint}) = LeftHalfOpenEndpoint(x)
+rightpoint(x::AbstractInterval, ::Type{<:HalfOpenEndpoint}) = RightHalfOpenEndpoint(x)
+
+function unbunch(interval::AbstractInterval, ::Type{T}; enumerate=false, sortby=identity) where T
+    if enumerate
+        return [(1, leftpoint(interval, T)), (1, rightpoint(interval, T))]
+    else
+        return [leftpoint(interval, T), rightpoint(interval, T)]
+    end
+end
+
+function unbunch(intervals, t::Type{T}=endpoint_type(intervals); enumerate=false,
+                 sortby=identity) where {T}
+    filtered = filter(i -> !isempty(intervals[i]), eachindex(intervals))
+    isempty(filtered) && return enumerate ? Tuple{Int, T}[] : T[]
+    if enumerate
+        result = mapreduce(((i) -> [(i, leftpoint(intervals[i], T)),
+                                    (i, rightpoint(intervals[i], T))]), vcat, filtered)
+        return sort!(convert(Array{Tuple{Int,T}}, result); by=sortby ∘ last)
+    else
+        result = mapreduce(((i) -> [leftpoint(intervals[i], T), rightpoint(intervals[i], T)]),
+                           vcat, filtered)
+        return sort!(convert(Array{T}, result); by=sortby)
+    end
+end
```

Comment:
If the point of a `HalfOpenEndpoint` is to produce a type-stable vector then why not just create two vectors where one contains only `LeftEndpoint`s and the other contains `RightEndpoint`s? This should also eliminate the need for the `enumerate` keyword. For sorting you'll need to use `sortperm` and apply the indices to both vectors.

The only disadvantage I see is that you could only sort on left or right endpoints but not both at the same time. Is that a use case you need for this though?

Reply:
> Is that a use case you need for this though?

Yup, that's exactly the goal to have all end points in order (so that e.g. interval A's endpoints may surround interval B's endpoints).

Reply:
Though, actually, looking at how I ended up doing this, I think that suggestion would work... 🤔

Reply:
Having worked through it a little more, I've simplified my approach but I still have a separate type of endpoint; I think this is the easiest to read / least friction. I wonder if it would be worth just briefly meeting and discussing this design choice.

The other option to simplify even more of this would be to remove the code that handles generic intervals, and just implement it for the Vector{<:Inteval{T,Open,Closed} and Vector{<:Inteval{T,Closed,Open} types. But that's exactly the case where the new `SimpleEndpoint` struct is useful, so it would not change that part.